File: netDxf/DxfDocument.Section.cs
using System.Collections.Generic;
using System.Linq;
using netDxf.Entities;
using netDxf.Header;
using netDxf.IO;
using netDxf.Objects;

namespace netDxf
{
    public sealed partial class DxfDocument
    {
        internal bool SectionReferencesRemoval(HashSet<DxfObject> removed)
        {
            // Ordinary collection removal cannot orphan a section's registered ownership graph.
            foreach (Section section in removed.OfType<Section>())
                if (this.Objects.Items.Any(item => DxfObjectDatabase.IsAncestor(section, item))) return true;
            foreach (DxfSectionSettings settings in this.Objects.Items.OfType<DxfSectionSettings>())
                if (!removed.Contains(settings) && settings.DatabaseReferences.Any(target => target != null && removed.Contains(target))) return true;
            var sections = new HashSet<DxfObject>(removed.OfType<Section>(), new MetadataIdentityComparer());
            if (sections.Count == 0) return false;
            if (this.Views.Any(view => !removed.Contains(view) && view.LiveSection != null && sections.Contains(view.LiveSection))) return true;
            foreach (DxfObject item in this.RetainedMetadataObjects())
            {
                if (removed.Contains(item)) continue;
                if (item.PersistentReactors.Any(sections.Contains)) return true;
                if (item is EntityObject entity && entity.Reactors.Any(sections.Contains)) return true;
                foreach (XData data in item.XData.Values) foreach (XDataRecord tag in data.XDataRecord)
                    if (tag.Code == XDataCode.DatabaseHandle && sections.Contains(this.GetObjectByHandle((string)tag.Value))) return true;
                if (item is DxfDatabaseObject databaseObject && databaseObject.DatabaseReferences.Any(sections.Contains)) return true;
                if (item is DxfXRecord record)
                    foreach (DxfTag tag in record.Data)
                        if (DxfObjectDatabase.IsReference(tag) && sections.Contains(this.GetObjectByHandle((string)tag.Value))) return true;
                if (item is DxfOpaqueObject opaque)
                    foreach (DxfTag tag in opaque.Tags)
                        if (tag.ValueType == DxfTagValueType.Handle && sections.Contains(this.GetObjectByHandle((string)tag.Value))) return true;
            }
            foreach (HeaderVariable variable in this.DrawingVariables.CustomValues())
            {
                DxfHandleKind kind = DxfGroupCode.GetHandleKind(variable.GroupCode);
                if (kind != DxfHandleKind.None && kind != DxfHandleKind.Arbitrary && variable.Value is string handle && sections.Contains(this.GetObjectByHandle(handle))) return true;
            }
            return false;
        }
    }
}

File: javascript/netDxf/DxfDocument.Section.js
// Copyright (c) netDxf contributors. MIT License; see package LICENSE.
import * as api from '../index.js';
export function InstallDocumentSection(Type) {
  Type.prototype.SectionReferencesRemoval = function(removed) {
    // The source initializes Objects even when no section is being removed.
    const objects = this.Objects;
    for (const section of removed) if (section instanceof api.Section && Array.from(objects.Items).some(item => api.DxfObjectDatabase.IsAncestor(section,item))) return true;
    for (const settings of objects.Items) if (settings instanceof api.DxfSectionSettings && !removed.has(settings) && Array.from(settings.DatabaseReferences).some(target => target !== null && removed.has(target))) return true;
    const sections = new Set(Array.from(removed).filter(item => item instanceof api.Section));
    if (sections.size === 0) return false;
    for (const view of this.Views) if (!removed.has(view) && view.LiveSection !== null && sections.has(view.LiveSection)) return true;
    for (const item of this.RetainedMetadataObjects()) {
      if (removed.has(item)) continue;
      if (Array.from(item.PersistentReactors).some(target => sections.has(target))) return true;
      if (item instanceof api.EntityObject && Array.from(item.Reactors).some(target => sections.has(target))) return true;
      for (const data of item.XData.Values) for (const tag of data.XDataRecord)
        if (tag.Code === api.XDataCode.DatabaseHandle && sections.has(this.GetObjectByHandle(tag.Value))) return true;
      if (item instanceof api.DxfDatabaseObject && Array.from(item.DatabaseReferences).some(target => sections.has(target))) return true;
      if (item instanceof api.DxfXRecord) for (const tag of item.Data)
        if (api.DxfObjectDatabase.IsReference(tag) && sections.has(this.GetObjectByHandle(tag.Value))) return true;
      if (item instanceof api.DxfOpaqueObject) for (const tag of item.Tags)
        if (tag.ValueType === api.DxfTagValueType.Handle && sections.has(this.GetObjectByHandle(tag.Value))) return true;
    }
    for (const variable of this.DrawingVariables.CustomValues()) {
      const kind = api.DxfGroupCode.GetHandleKind(variable.GroupCode);
      const value = variable.Value instanceof api.BoxedString ? variable.Value.Value : variable.Value;
      if (kind !== api.DxfHandleKind.None && kind !== api.DxfHandleKind.Arbitrary && typeof value === 'string' && sections.has(this.GetObjectByHandle(value))) return true;
    }
    return false;
  };
}
